test/common/TestGradient.hs

```diff
@@ -288,28 +290,57 @@ testBarR =
 
 -- The following are borrowed from https://github.com/benl23x5/adops.
 
--- | Derivative of unpadded full convolution with respect to the input image.
+-- | Derivative of padded full convolution with respect to the input image.
 --
 conv2d_dInp
-  :: forall shK shA shB shK1 nImgs nCinpA nAh nAw nCoutK nCinpK nKh nKw r.
+  :: forall shK shA shB shB1 sh1 nAimg nAcha nAh nAw nKimg nKcha nKh nKw r.
```

Comment:
For our sanity, could we stick to the same dimension size names here and in DualNumber.hs? In DualNumber.hs, in turn, I tried to stick to the dimension names from

https://github.com/benl23x5/adops/blob/8c16600e58e1b4c8181daa6a28f0d9765c2f9d48/src/Adops/Op/Conv.hs#L6

as opposed to 

https://github.com/benl23x5/adops/blob/8c16600e58e1b4c8181daa6a28f0d9765c2f9d48/src/Adops/Op/Conv.hs#L124

but I'm fine with either choice, as long as it's consistent.

Reply:
ok I'll make them consistent

Reply:
Thank you. That also makes it easier for me to spot what I don't understand. E.g., is it significant that in `conv2d_dInp` there is, in https://github.com/benl23x5/adops/blob/master/src/Adops/Op/Conv.hs

    Shape4 nImgs  nCoutB nBh nBw = shape arrB

but in `TestGradient.hs`

    shB  ~ '[nImgs, nCout, nAh, nAw]

(where the latter agrees with  `conv2d` in `DualNumber.hs` and in `Conv.hs`). Or am I badly pattern-matching this? Or is naming in horde-ad code made consistent only internally, not wrt adops code?

Reply:
In Adops the shape is given in value-land. So we have a runtime check:
```
 = let  Shape4 nImgs  nCoutB nBh nBw = shape arrB
        Shape4 nCoutK nKchan nKh nKw = shape arrK
        nCout   = same nCoutB nCoutK -- blow up if no good
```
In horde-ad the shape is encoded in the type. To match Adops syntactically we could write:
```
forall nCoutB nCoutK....
shB  ~ '[nImgs, nCoutB, nAh, nAw],
shK  ~ '[nCoutK, nKchan, nKh, nKw],
nCoutB ~ nCoutK
```
or just use the same name `nCout` and get the same constraint result.

Reply:
Yes, that makes sense; let's keep `nCout` as is. However, I was alluding to there being "nBh nBw" in the first case vs "nAh, nAw" in the second.

Reply:
oh right, sorry I missed that. In this particular version of convolution we are padding the input such that the output's spatial dimensions are the same as the input's. Having it in the types is a nice bonus for horde!

Reply:
Oh, right, now I can see "the output size is the same as the input size", so that's in the comment, not in the variable names, while type-level requires it to be reflected everywhere. Puzzle solved.

BTW, could you kindly `git rebase master`? Then I'd merge immediately, before it gets out of sync again.